Allow OSD driven parameter configuration

andyp1per opened the pull request:
This PR adds two new "parameter configuration" OSD screens.

Any parameter can be put on a screen and transmitter roll and pitch can be used to cycle through the parameters, selecting one to change and then using pitch up and down to raise or lower the value.

While modifying parameters arming is disabled.
If the parameter screen is selected while armed then modification is disabled

peterbarker commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -190,10 +233,90 @@ class AP_OSD_Screen
+public:
+    // configured index.
+    AP_Int16 index;
+    // parameter number
```
Include this in the variable name....

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -190,10 +233,90 @@ class AP_OSD_Screen
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+    AP_OSD_ParamSetting params[NUM_PARAMS] = {
```
Shouldn't this be `static const` somehow?

andyp1per replied:
These are just the initial defaults. I'm not sure how I would make these static without using up more memory

peterbarker commented on libraries/AP_OSD/AP_OSD_ParamScreen.cpp:
```diff
@@ -0,0 +1,449 @@
+extern const AP_HAL::HAL& hal;
+
+// constructor
+AP_OSD_ParamScreen::AP_OSD_ParamScreen()
```
Omit this?

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_OSD/AP_OSD_ParamScreen.cpp:
```diff
@@ -0,0 +1,449 @@
+            param->set(vec);
+        }
+            break;
+        default:
```
Omit the default and just break-and-do-nothing for types you don't understand.  There's not that many :-)

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_OSD/AP_OSD_ParamScreen.cpp:
```diff
@@ -0,0 +1,449 @@
+{
+    // currently in the OSD menu, do not allow arming
+    if (!is_arming_allowed()) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "In OSD menu");
```
This is making an assumption about why `is_arming_allowed()` said no.

andyp1per replied:
Fixed the name

peterbarker commented on libraries/AP_OSD/AP_OSD_Setting.cpp:
```diff
@@ -48,10 +48,95 @@ const AP_Param::GroupInfo AP_OSD_Setting::var_info[] = {
 // constructor
 AP_OSD_Setting::AP_OSD_Setting(bool _enabled, uint8_t x, uint8_t y)
 {
     enabled = _enabled;
```
Usually you'd want to do these assignments as part of the signaturey bit rather than the body here.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_OSD/AP_OSD_Setting.cpp:
```diff
@@ -48,10 +48,95 @@ const AP_Param::GroupInfo AP_OSD_Setting::var_info[] = {
+AP_OSD_ParamSetting::AP_OSD_ParamSetting(uint8_t param_number, bool _enabled, uint8_t x, uint8_t y, uint16_t idx)
+    : AP_OSD_Setting(_enabled, x, y)
+{
+    index = idx;
```
As above, usually this goes after the `:` in the definition.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_OSD/AP_OSD_Setting.cpp:
```diff
@@ -48,10 +48,95 @@ const AP_Param::GroupInfo AP_OSD_Setting::var_info[] = {
+{
+    _number = param_number;
+    index = -1;
+    strcpy(_name, name);
```
`strcpy` is a yellow card offence :-P

andyp1per replied:
strncpy now

peterbarker commented on libraries/AP_OSD/AP_OSD_Setting.cpp:
```diff
@@ -48,10 +48,95 @@ const AP_Param::GroupInfo AP_OSD_Setting::var_info[] = {
+        enabled = false;
+    } else {
+        param->copy_name_token(token, _name, 16);
+        _name[16] = 0;
```
This doesn't make a lot of sense - what if _name is short?  Does `copy_name_taken` null-terminate?

andyp1per replied:
It seems to, except if its the full width. Not sure what else I can do?

andyp1per commented:
@peterbarker you wanted a video, so here you go https://youtu.be/nDpsjplFSI0

andyp1per commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -913,6 +914,26 @@ bool AP_Arming::camera_checks(bool display_failure)
 
+bool AP_Arming::osd_checks(bool display_failure)
+{
+    if (checks_to_perform & ARMING_CHECK_ALL) {
```
Fixed

andyp1per commented on libraries/AP_Arming/AP_Arming.h:
```diff
@@ -163,6 +163,8 @@ class AP_Arming {
 
     bool camera_checks(bool display_failure);
 
+    bool osd_checks(bool display_failure);
```
Yes.

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -256,8 +256,16 @@ bool AP_RunCam::pre_arm_check(char *failure_msg, const uint8_t failure_msg_len)
 // OSD update loop
 void AP_RunCam::update_osd()
 {
+    bool armed = AP::arming().is_armed();
```
Renamed, but can't be const

andyp1per commented on libraries/AP_OSD/AP_OSD.cpp:
```diff
@@ -300,26 +315,26 @@ void AP_OSD::update_current_screen()
     // Switch on failsafe event
     if (AP_Notify::flags.failsafe_radio || AP_Notify::flags.failsafe_battery) {
-        if (!was_failsafe && failsafe_scr > 0 && failsafe_scr <= AP_OSD_NUM_SCREENS && screen[failsafe_scr-1].enabled) {
+        if (!was_failsafe && failsafe_scr > 0 && failsafe_scr <= AP_OSD_NUM_SCREENS && get_screen(failsafe_scr-1).enabled) {
```
No, its used as the raw index in many places

andyp1per commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -47,30 +80,55 @@ class AP_OSD_Setting
+protected:
+    bool check_option(uint32_t option);
+
+    enum unit_type {
```
This is old code. I only moved it around. Would prefer not to clean up.

Hwurzburg commented:
This is a super addition!....the stick selection method seems pretty standard...
but It appears that this is hard coded for parameter lists...which would be impossible for most users....
I have used RVSOD, BF, and VeEctor OSD param adjustment screens for many years....I would suggest a system for parameter setup much like Alex's BiDir Frsky system for TXs which is extremely easy to use and setup.... https://github.com/yaapu/FrskyLuaGCS... I have done a lot of testing and the wiki for it...its beautiful...
I know we don't want to mess with the SD for storage, but we do for Scripts and using it for pre-determined param screens is a tiny tiny bit of storage...you could easily adapt Alex's Window GUI to produce the .lua files containing the params, min,max, and increment....then its easily setup and changed by the average user...we could even integrate MAVftp directly into the Window GUI for SD transfer or use MP MAVftp when its able to access the SD file system for script upload

andyp1per commented:
Thanks for the feedback @Hwurzburg - do you have any example screens? The issue with scripting is it won't work on 1mb boards, so I think we need a simple solution for those setups. I agree a script based solution would be desirable in addition since it would mean users could configure the screens themselves. I considered doing a menu based solution, but that quickly gets very complicated (see my runcam PR) and figured starting with something simple was a good addition.

What do you mean by "hardcoded for parameter lists"? Are there other things you want to play with apart from parameters?

Hwurzburg commented:
I think you will get a lot of ideas by playing with Alex's stuff

andyp1per commented:
Just to clarify based on our conversation:
- There is no hard-coding. Any combination of parameters can be selected for display on a screen. This would need to be done through some config tool such as mission planner, which does this today for the read-only screens
- The manipulation is only currently as good as AP_Param allows. None of the metadata is available in the firmware so we currently cant set limits etc. Storing some metadata would allow an improved user experience but at a flash cost.
- The parameters currently selected are ones I deemed useful (PID tuning values). They can be changed by the user and different defaults could be selected.
- 1Mb boards do not have scripting, so scripting is not an option here - although it could be used to provide an extended user experience if we wanted

Hwurzburg commented:
thanks....nice work...my concerns were easy user selection of params to display and ease of use at the field....but  still unsure how to have the ability to increment things like cdeg which needs fast adjustment of units (by 10s?) versus pids which may need adjustment by 0.01 increments... expo stick rates can help but not sure that some kind of increment value doesn't need to be stored per parm index for the adjustments..maybe you have that covered...

andyp1per commented:
Yes I agree *some* meta data needs to be stored to make the UX acceptable, that's why it would be great to have a conversation about it on the dev call.

vierfuffzig commented:
@andyp1per just tested on SITL / plane. this is absolutely awesome, thanks a lot!

andyp1per commented:
@vierfuffzig glad you like it! We had a long discussion on the dev call about what the UX should be. On the one hand it needs to be usable (so understandable interface, stops the user shooting themselves in the foot) but on the other hand it needs to be simple enough to fit on 1Mb boards and be flexible enough to be configurable to user requirements. I have added man/max/incr to allow the user to constrain how the values get modified, but I'm not sure whether the UX can be improved much further without introducing a lot of complexity. Glad to know people's thoughts. Some ideas:

- Allow regular OSD screens to have one parameter value that can be modified by Ch6
- Have support for parameter groups, such that selecting a group goes into a submenu. This gets really complicated really quickly however
- Abbreviate the parameters names somehow for less flash cost and more screen real estate
- Have a table of min/max/incr for well-known parameter values so that the user doesn't have to configure them (could also be scraped by MP from the docs if they exist)

vierfuffzig commented:
@andyp1per i've played around a bit, trying to figure the plane params i'd consider most useful on a screen. i think we can offer some vehicle-specific top 8 defaults, but in the end i feel like this will always be asking for full param access support really.
i had something like this in mind, don't know if it's doable:

```
Param: [name, full list alphabetically scrollable]
Val: [n.nn, settable in incr as set below]
Increment: [0.001 | 0.01 | 0.1 | 1 | 10 , selectable]
```

plus optionally:

`Reset to default: [full right for 3s to reset]`

- i personally would keep param access restricted to dedicated screens
- i doubt param groups will help a lot without a sheet at hand to look them up
- imho min / max only add safety if preconfigured and hardcoded. i personally am not a fan. risk of accidentally changing a param to out of range value is quite limited on this system.

andyp1per commented:
It's all doable just a question of complexity, usability and storage. 

I'm not a fan of displaying the whole list - I think that's really unwieldy. What I did think is you could have full left for 3s say which then allowed you to cycle through the parameters themselves rather than the values.

I agree that hardcoded min/max would be better. I suppose we could do this for the most used parameters. It also might be possible to do a reasonable encoding to save space - for instance max digits from -8 to +8 could be encoded in a nibble

vierfuffzig commented:
> cycle through the parameters themselves

+1, that's what i meant

Hwurzburg commented:
re min/max I think that its a requirement myself, it does require thought/study before setup, but many params can get you into trouble quick if out of range, and on the other hand many times you may want to be able push one out of range....I have many times setting up yaw on imperfectly aligned quadplanes...

also, how is something like ARMING_CHECK handled?...a param I use a lot in test and troubleshooting flights....a pure integer is really only workable for 0 and 1 cases in the present scheme I think...and its 16bits ....a lot of numbers to scroll thru...

@vierfuffzig Basti, if you have an Frsky setup, you should try Alex's tx based param editor... https://github.com/yaapu/FrskyLuaGCS/wiki  its close to finishing up...just fyi

andyp1per commented:
@Hwurzburg I really don't think checkboxes are going to be workable through stick control except for very limited ranges of options (e.g. a single select bitmask). I don't think there are even the OSD icons required. So quite a lot of effort and complexity for very little value. Just my $0.02 :)

Hwurzburg commented:
I agree...just wondering how they are handled now

andyp1per commented:
I've made the change to use the ParamToken. It works well and I can generate a list of all of the values together with the parameter names and plug them into the OSD and get the parameter I want. So seems like a reasonable approach. I'll have to do the same for plane and see if I get any differences.
Since we now have the token I will remove the caching of parameter names since we can get those directly easily enough, it will reduce flash usage by 17*8*2 bytes.

vierfuffzig commented:
@andyp1per one issue i noticed in current config: if you don't actively set a chosen param's min / max / incr to useful values other than 0, you can well set the param value to 0 accidentally, but not change it back again via OSD. +1 for hardcoded imho.

andyp1per commented:
@vierfuffzig this is partially implemented now and partially working. Still not quite there but coming along nicely

vierfuffzig commented:
@andyp1per this is really awesome, fantastic work!

andyp1per commented:
> 
> 
> @andyp1per this is really awesome, fantastic work!
> 
> ![paramscrn](https://user-images.githubusercontent.com/9342253/81843048-56d45d80-954d-11ea-8508-fd331b50add0.jpeg)

Very cool!

andyp1per commented:
Symbolic names working for the most part:

![image](https://user-images.githubusercontent.com/2893260/81869951-db82a400-956c-11ea-980c-fad5532545d8.png)

andyp1per commented:
Symbolic names all working and added a save button. I think this is pretty much where I want it to be:

![image](https://user-images.githubusercontent.com/2893260/81984809-45fb1900-962d-11ea-8187-db3c9bda73ce.png)

vierfuffzig commented:
@andyp1per great work again, thanks a lot! do static labels work on copter only (didn_t get those on plane)? and what do you think about sorting tokens alphabetically by param name?

andyp1per commented:
hmmn, maybe that means the indexes are different on plane - which would be a pain! I'll have to investigate.
Sorting would require that we store the index of the actual item and set the parameter to that. It's doable, just not sure how much it's worth it. I'm also not sure we want all possible labels. Another approach would be a smaller set of labels for most frequently used things that we flip through. But then I have no idea what's popular - and this pushes against being generic enough to cover all parameters.

vierfuffzig commented:
thanks again! needs adding specific keys to AP_OSD_ParamSetting::ParamMetadata to make labels work on plane:

```
const AP_OSD_ParamSetting::ParamMetadata AP_OSD_ParamSetting::_param_metadata[] = {
    // copter-specific param keys
    { PARAM_INDEX(6, 0, 11), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES }, // SERIAL_PROTOCOL0
    { PARAM_INDEX(6, 0, 1), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // SERIAL_PROTOCOL1
    { PARAM_INDEX(6, 0, 3), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // SERIAL_PROTOCOL2
    { PARAM_INDEX(6, 0, 5), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // SERIAL_PROTOCOL3
    { PARAM_INDEX(6, 0, 7), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // SERIAL_PROTOCOL4
    { PARAM_INDEX(6, 0, 9), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // SERIAL_PROTOCOL5
    { PARAM_INDEX(6, 0, 12), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES }, // SERIAL_PROTOCOL6
    { PARAM_INDEX(100, 0, 24657), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC1_OPTION
    { PARAM_INDEX(100, 0, 24721), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC2_OPTION
    { PARAM_INDEX(100, 0, 24785), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC3_OPTION
    { PARAM_INDEX(100, 0, 24849), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC4_OPTION
    { PARAM_INDEX(100, 0, 24913), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC5_OPTION
    { PARAM_INDEX(100, 0, 24977), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC6_OPTION
    { PARAM_INDEX(100, 0, 25041), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC7_OPTION
    { PARAM_INDEX(100, 0, 25105), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC8_OPTION
    
    // plane-specific param keys
    { PARAM_INDEX(3, 0, 11), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES }, // SERIAL_PROTOCOL0
    { PARAM_INDEX(3, 0, 1), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // SERIAL_PROTOCOL1
    { PARAM_INDEX(3, 0, 3), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // SERIAL_PROTOCOL2
    { PARAM_INDEX(3, 0, 5), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // SERIAL_PROTOCOL3
    { PARAM_INDEX(3, 0, 7), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // SERIAL_PROTOCOL4
    { PARAM_INDEX(3, 0, 9), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // SERIAL_PROTOCOL5
    { PARAM_INDEX(3, 0, 12), -1, 28, 1, ParamMetadata::StringValues, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES }, // SERIAL_PROTOCOL6
    { PARAM_INDEX(150, 0, 24657), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC1_OPTION
    { PARAM_INDEX(150, 0, 24721), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC2_OPTION
    { PARAM_INDEX(150, 0, 24785), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC3_OPTION
    { PARAM_INDEX(150, 0, 24849), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC4_OPTION
    { PARAM_INDEX(150, 0, 24913), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC5_OPTION
    { PARAM_INDEX(150, 0, 24977), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC6_OPTION
    { PARAM_INDEX(150, 0, 25041), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },  // RC7_OPTION
    { PARAM_INDEX(150, 0, 25105), 0, 104, 1, ParamMetadata::StringValues, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS }   // RC8_OPTION
};
```

andyp1per commented:
Cool, as it's only the keys that are different, that's pretty easy to fix with a #define. Sometimes low tech is the best :)

vierfuffzig commented:
might be worth considering statics for baudrates, flightmodes and servo functions too. the latter are quite a few though...

andyp1per commented:
Yes agreed and failsafe actions. My plan was to perhaps only enable these for 2Mb boards. I guess it would also be possible to use a simple compression scheme since these are a very limited character set (uppercase, space and _) pretty much, so could probably get compression by a factor of 20 or so.

andyp1per commented:
@vierfuffzig feel free to PR against this with additional fixed lists and I will incorporate them.

vierfuffzig commented:
@andyp1per thanks, will do! have some labels ready and tested but then unexpectedly flyable weather interfered...

vierfuffzig commented:
@andyp1per spent quite some time trying to do more static labels with a pretty steep learning curve. i learned we not only need per-vehicle param groups and keys, but individual labels too in most cases. then learned there might even be per-vehicle group/key mismatches due to different features set in compile-time plus hardware tokens don‘t necessarily match SITL.
setting set KEY_DUMP i got fatal build errors on printf format conversion, managed to solve them finally. 
now i can‘t figure where to find the actual key dump output, appreciate your help here.

andyp1per commented:
To see the key dump output use mavproxy to display.

Hmmn, so sounds like we really do need some kind of compile time detection of the keys.

vierfuffzig commented:
@andyp1per thanks for your guidance, finally managed to retrieve a dump from actual hardware, not a complete one though, as it gets interrupted by other console prompts. enough to see a pattern in group/token shift as compared to SITL. small steps...
![osd tokens plane matek wing](https://user-images.githubusercontent.com/9342253/83811876-e091c980-a6ba-11ea-8e40-a5d43f65e0d5.jpeg)

andyp1per commented:
Very cool!

andyp1per commented:
Squashed and cleaned up the merge. Probably ready for a review.

vierfuffzig commented:
@andyp1per tested on MatekF405-Wing and MatekF765-Wing. unfortunately param keys diverge with varying build features:
[F405-Wing tokens](https://gist.github.com/vierfuffzig/9622fd04e2be3ee4fa7bf216e471274c)
[F765-Wing tokens](https://gist.github.com/vierfuffzig/f4449b857b9def2b39e901df17b9ac84)
param identification by key likely requires per-board defines and noticeable effort to keep aligned with feature additions.

andyp1per commented:
That's disappointing. Scratching my head over what to do.

vierfuffzig commented:
@andyp1per checked PKeys 405 vs 765 plane, looking absolutely promising! will redo labels and PR asap.

vierfuffzig commented:
@andyp1per sry, now realized you did the new keys already... thanks a lot again!
is there a reasonable way to lable params with large increment gaps like baudrates?

andyp1per commented:
Flight test was ok. Couple of glitches I would like to fix:

- Roll hard left for 3s was not working
- Up and down actions are still a little glitchy, need to figure out a better de-bounce strategy
- Might be useful to have a reboot button as well as save since some parameters need that
- Very small floats are not displaying correctly, so for instance 0.00037 is displaying as 0.0003

andyp1per commented:
Ok glitches fixed. 
The save button is context sensitive - if parameters need saving it says save otherwise reboot
Reboot works on hold right on save for 3s
Small floats handled
Added failsafe options

andyp1per commented:
Working well for me on my 3" gecko but would like your feedback @vierfuffzig

vierfuffzig commented:
@andyp1per i‘ve been using this on a couple of setups since first commits, with good results. for someone like me who doesn‘t use bidir telemetry this adds major convenience.
up/down indeed needed some attention, well manageable though. will test new debounce asap.
the save button is below PAL borders iirc, only visible on NTSC. might be favourable to move it further up (3 lines).
probably a per-vehicle choice of preselected parameters will be helpful. 
parameter order imho is the primary limitation. trying to find a parameter by scrolling can be unpleasant. the list follows vehicle/Parameters.cpp order, doesn‘t it...?

vierfuffzig commented:
@andyp1per major improvement in up/dn, awesome! 
SAVE button changes to REBOOT after use on all parameters now. my perception was it only should on @RebootRequired?

andyp1per commented:
There's no way to access @RebootRequired - it's not available, we would have to configure it for every parameter.
Another option is to maybe make it two step - so hold right 3s SAVE->REBOOT, hold right another 3s actually reboot.
I can make the save location configurable.

vierfuffzig commented:
@andyp1per ok thanks, so user error on my side for the reboot feature. no functional changes required imho.
+1 for adding save button NTSC option or _X and _Y.
perspectively UI integration (MP OSD tab) for picking and placing items would round off user experience.
keeping static lables aligned with option addons will likely need some care...

andyp1per commented:
Ok, made the save location configurable. This also revealed a bug with parameter defaults which I have fixed.

andyp1per commented:
I've done another video
https://www.youtube.com/watch?v=Ejhc0VCrf4c

peterbarker commented on Tools/ardupilotwaf/cxx_checks.py:
```diff
@@ -257,6 +257,9 @@ def check_SFML(cfg, env):
         cfg.msg("Checking for SFML graphics:", 'disabled', color='YELLOW')
         return False
+    if cfg.env.DEST_OS == 'darwin':
+        cfg.env.append_value('INCLUDES', ['/usr/local/include'])
```
This is rather odd.

Presumably SFML has been installed from a non-packaged source?

If so - should we get some documentation together?

andyp1per replied:
The easiest way to install SFML on macos is to use homebrew - that's what gives this location.

BUT it doesn't work in SITL because there is a macos limitation in SFML that prevents you doing any video operations from anywhere but the main thread. So to make this work requires some retooling of the sitl backend.

peterbarker replied:
So should these lines be removed as causing noise ATM?

andyp1per replied:
Removed

peterbarker commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -20,14 +20,50 @@
+#define AP_OSD_NUM_PARAM_SCREENS 2
+#define AP_OSD_NUM_SCREENS (AP_OSD_NUM_DISPLAY_SCREENS + AP_OSD_NUM_PARAM_SCREENS)
+
+#define SYM_M           0xB9
```
All of these defines are going to be present in everything which includes `AP_OSD.h`.  Could we look at moving these into the implementation, or as static-consts in the object?

andyp1per replied:
moved into implementation

sh4jonny commented:
💪👍

peterbarker commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -925,6 +926,26 @@ bool AP_Arming::camera_checks(bool display_failure)
 
+bool AP_Arming::osd_checks(bool display_failure) const
+{
+    if ((checks_to_perform & ARMING_CHECK_ALL) || (checks_to_perform & ARMING_CHECK_CAMERA)) {
```
Is `CAMERA` a good fit here?

There's nothing better?

andyp1per replied:
I think CAMERA was your suggestion on a previous review 😄 

It depends how granular we want to get. I think camera is ok unless you want a specific OSD one.

peterbarker commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -278,6 +456,14 @@ class AP_OSD
+    }
+
+    // Check whether arming is allowed
+    bool pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) const;
```
We should really canonicalise on one of `pre_arm_check` or `pre_arm_checks`

andyp1per replied:
Which one do you prefer? They seem pretty evenly distributed through the code. Maybe slightly more pre_arm_check()s

peterbarker commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -190,10 +236,140 @@ class AP_OSD_Screen
+ */
+class AP_OSD_ParamSetting : public AP_OSD_Setting
+{
+public:
```
Please check visibility on members on this class.  Pretty sure some of them can be private.

andyp1per replied:
I already tried that, pretty sure they all have to be public unless we make this class a friend of AP_OSD_ParamScreen, which seems like its defeating the point.

andyp1per replied:
I tell a lie - _metadata_index can be private, but that is all

andyp1per commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -925,6 +926,26 @@ bool AP_Arming::camera_checks(bool display_failure)
+bool AP_Arming::osd_checks(bool display_failure) const
+{
+    if ((checks_to_perform & ARMING_CHECK_ALL) || (checks_to_perform & ARMING_CHECK_CAMERA)) {
+#if OSD_ENABLED
```
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -257,8 +257,16 @@ bool AP_RunCam::pre_arm_check(char *failure_msg, const uint8_t failure_msg_len)
+    bool use_armed_loop = AP::arming().is_armed();
+#if OSD_ENABLED
+    // prevent runcam stick gestures interferring with osd stick gestures
+    AP_OSD* osd = AP::osd();
```
Could be `const`.

peterbarker replied:
We only need to do all of this work if use_armed_loop isn't already set to true...

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_OSD/AP_OSD_ParamScreen.cpp:
```diff
@@ -0,0 +1,663 @@
+}
+
+// return radio values as LOW, MIDDLE, HIGH
+RC_Channel::AuxSwitchPos AP_OSD_ParamScreen::get_channel_pos(uint8_t rcmapchan) const
```
We have an equivalent function on RC_Channel now.

andyp1per replied:
They are not quite the same and it didn't seem worth refactoring to pass in a bunch of different parameters for my use case.

peterbarker replied:
Please provide a really big comment here explaining that the ranges are different to the RC_Channel library, and why.

peterbarker commented on libraries/AP_OSD/AP_OSD_ParamScreen.cpp:
```diff
@@ -0,0 +1,663 @@
+
+    const Event ev = map_rc_input_to_event();
+    // only take action on transitions
+    if (ev == Event::NONE && ev == _last_rc_event) {
```
This looks... odd....

andyp1per replied:
You only skip this step when you get None twice in a row.

peterbarker commented on libraries/AP_Param/AP_Param.cpp:
```diff
@@ -2519,16 +2519,16 @@ void AP_Param::show(const AP_Param *ap, const char *s,
 {
     switch (type) {
     case AP_PARAM_INT8:
-        port->printf("%s: %d\n", s, (int)((AP_Int8 *)ap)->get());
```
Should this be pulled out into a separate PR?

andyp1per replied:
It's fairly fundamental to the mapping I'm using

tridge commented:
discussed on EU dev call, decided to do a mavlink2 msg for configuring OSD parameters, example here:
https://github.com/tridge/mavlink/commits/osd-param-andy

andyp1per commented:
Ok, reworked to use mavlink and MAVProxy to set OSD parameter types. To build you also need to use https://github.com/ArduPilot/MAVProxy/pull/800 and https://github.com/ArduPilot/mavlink/pull/147 so it's somewhat involved.
@vierfuffzig I'd be grateful if you could give this a go, but understand that setup might be a little tricky

vierfuffzig commented:
@andyp1per i‘ll sure do, thanks a lot! on a sidenote i’m using the param screen on all my setups for quite a while now with excellent results, it’s exceptionally convenient when dialing in new setups, setting throttle ranges, loiter radii, checking servo auto trims etc...

andyp1per commented:
@vierfuffzig cool! So you'll see in https://github.com/ArduPilot/MAVProxy/pull/800 that the new mavproxy osd module allows you to set these things conveniently from the command line.

vierfuffzig commented:
@andyp1per 
i'm struggling a bit with the mavproxy module, getting error messages on module load:
![image](https://user-images.githubusercontent.com/9342253/90808494-615a7c00-e320-11ea-9a82-2bd85bbfb7f8.png)
likely user error, appreciate your help...

andyp1per commented:
You are missing the mavlink bindings. You need to build in pymavlink and then mavproxy using the mavlink bindings. You have to set MDEFS. I’ll try and send the magic @peterbarker incantation later

Sent from my iPhone

> On 20 Aug 2020, at 20:08, vierfuffzig <notifications@github.com> wrote:
> 
> ﻿
> @andyp1per
> i'm struggling a bit with the mavproxy module, getting error messages on module load:
> 
> likely user error, appreciate your help...
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

andyp1per commented:
@vierfuffzig I run this in both pymavlink and mavproxy:
```
MDEF=/Users/andy/github/ardupilot-dev/modules/mavlink/message_definitions python setup.py build install --user
```
I had to uninstall both completely before this would work properly (using pip uninstall)

vierfuffzig commented:
thanks andy, uninstalling mavproxy and reinstalling it after pymav MDEF did the trick.

i got it basically working, showing and setting params. i see argument count errors when trying to set params with min max increment and i do see arrays randomly applied to wrong or non-array params, like finding servo functions applied to SERIALn_PROTOCOL or unrelated parameters:

![image](https://user-images.githubusercontent.com/9342253/90824433-b5248f80-e337-11ea-9251-df920d96aa7d.png)

![image](https://user-images.githubusercontent.com/9342253/90824565-e56c2e00-e337-11ea-95a0-4ff523b6517c.png)

regarding my particularly limited experience with the pymav environment, i still can't exclude user- / install errors obviously...

andyp1per commented:
The min/max thing must be a bug - poor testing on my part. I will write some auto tests when I get the thumbs up from @tridge and @peterbarker for the approach.
There are no restrictions on what parameter type you can apply the mnemonics to - so you could apply AUX_OPTIONS to a SERIALn_PROTOCOL. When you say "random" can you be a little more specific? The string names are these BTW:
```
        self.type_map = {
            mavutil.mavlink.OSD_PARAM_NONE : "NONE",
            mavutil.mavlink.OSD_PARAM_SERIAL_PROTOCOL : "SERIAL_PROTOCOL",
            mavutil.mavlink.OSD_PARAM_SERVO_FUNCTION : "SERVO_FUNCTION",
            mavutil.mavlink.OSD_PARAM_AUX_FUNCTION : "AUX_FUNCTION",
            mavutil.mavlink.OSD_PARAM_FLIGHT_MODE : "FLIGHT_MODE",
            mavutil.mavlink.OSD_PARAM_FAILSAFE_ACTION : "FAILSAFE_ACTION",
            mavutil.mavlink.OSD_PARAM_FAILSAFE_ACTION_1 : "FAILSAFE_ACTION_1",
            mavutil.mavlink.OSD_PARAM_FAILSAFE_ACTION_2 : "FAILSAFE_ACTION_2" }
```

vierfuffzig commented:
thanks! see screen shot above. i did 
`osd param-set 5 1 LOG_FILE_BUFSIZE`
and randomly got SERIAL_PROTOCOL mnemonics applied. that slot was set to RLL2SRV_P previously btw.
as well as SERIAL1_PROTOCOL on slot 6 showing SERVO_FUNCTIONS items.
i didn‘t specifiy the matching string though.

andyp1per commented:
@vierfuffzig - tested fine on my mini quad as well. Just had to add the new protocols that have since been added.

vierfuffzig commented:
@andyp1per formidable, aweseome work! 
maintaning type arrrays will likely need some attention. and i‘d expect questions about GUI integration perspectively ;)

tridge commented on libraries/AP_OSD/AP_OSD_ParamScreen.cpp:
```diff
@@ -0,0 +1,700 @@
+            param->set(constrain_float(param->get() + incr, setting._param_min, setting._param_max));
+        }
+            break;
+        case AP_PARAM_VECTOR3F:
```
need to test if vectors do really work

andyp1per replied:
Removed

tridge commented on libraries/AP_OSD/AP_OSD_ParamScreen.cpp:
```diff
@@ -0,0 +1,700 @@
+        if (params[i].enabled) {
+            AP_Param* p = params[i]._param;
+            if (p != nullptr) {
+                p->save();
```
try to avoid the save if possible?

tridge commented on libraries/AP_OSD/AP_OSD_ParamScreen.cpp:
```diff
@@ -0,0 +1,700 @@
+    }
+
+    const uint16_t in = chan->get_radio_in();
+    if (in <= 900 or in >= 2200) {
```
use || instead of 'or'

andyp1per replied:
Did not even realise I had done this!

andyp1per replied:
Copied from RC_Channel.cpp:1096 so I blame @peterbarker 😉

tridge commented on libraries/AP_OSD/AP_OSD_ParamSetting.cpp:
```diff
@@ -0,0 +1,418 @@
+// copter parameters
+const AP_OSD_ParamSetting::ParamMetadata AP_OSD_ParamSetting::_param_metadata[OSD_PARAM_NUM_TYPES] = {
+    { -1, 31, 1, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // OSD_PARAM_SERIAL_PROTOCOL
+    { 0, 132, 1, ARRAY_SIZE(SERVO_FUNCTIONS), SERVO_FUNCTIONS },                // OSD_PARAM_SERVO_FUNCTION
```
use k_nr_aux_servo_functions ?

tridge commented on libraries/AP_OSD/AP_OSD_ParamSetting.cpp:
```diff
@@ -0,0 +1,418 @@
+
+// copter parameters
+const AP_OSD_ParamSetting::ParamMetadata AP_OSD_ParamSetting::_param_metadata[OSD_PARAM_NUM_TYPES] = {
+    { -1, 31, 1, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // OSD_PARAM_SERIAL_PROTOCOL
```
need to add a #define in AP_SerialManager.h

tridge commented on libraries/AP_OSD/AP_OSD_ParamSetting.cpp:
```diff
@@ -0,0 +1,418 @@
+    }
+
+    if (_param == nullptr) {
+        hal.console->printf("Couldn't find param for key: %d, idx: %d, group: %d\n", int(_param_key.get()), int(_param_idx.get()), int(_param_group.get()));
```
need to remove, or change to send_text with rate limit

tridge commented on libraries/AP_OSD/AP_OSD_ParamSetting.cpp:
```diff
@@ -0,0 +1,418 @@
+    // if a parameter was configured then use that
+    _current_token = AP_Param::ParamToken {};
+    // surely there is a more efficient way than brute-force search
+    for (_param = AP_Param::first(&_current_token, &_param_type);
```
ok with cpu cost? only called when user makes a change?

andyp1per replied:
Yes cached and only changed when the user changes it

tridge commented on libraries/AP_OSD/AP_OSD_ParamSetting.cpp:
```diff
@@ -0,0 +1,418 @@
+            break;
+        }
+
+        if (!_param_min.configured() || force) {
```
configured() is expensive, need to make sure not being called at high rate

andyp1per replied:
only called when the user has made a change

tridge commented on libraries/AP_OSD/AP_OSD_ParamSetting.cpp:
```diff
@@ -0,0 +1,418 @@
+    // is no longer valid so reset these to guessed values
+    guess_ranges(true);
+    if (_param_min.configured()) {
+        _param_min.save();
```
prefer to avoid save() if we can to save storage

andyp1per replied:
Bitmask used to check whether parameter has been modified

tridge commented on libraries/AP_OSD/fonts/Param-Tokens-Plane.md:
```diff
@@ -0,0 +1,2115 @@
+|Key|PKey|Index|Group|Name|Value|
```
this will get out of date pretty quickly, better to not include I think

andyp1per replied:
Removed

tridge commented on libraries/AP_OSD/fonts/Parameter-Tokens.md:
```diff
@@ -0,0 +1,1025 @@
+|Key|PKey|Index|Group|Name|Value|
```
remove this too

andyp1per replied:
Removed

tridge commented on libraries/AP_Param/AP_Param.cpp:
```diff
@@ -2519,16 +2539,16 @@ void AP_Param::show(const AP_Param *ap, const char *s,
     switch (type) {
     case AP_PARAM_INT8:
-        port->printf("%s: %d\n", s, (int)((AP_Int8 *)ap)->get());
+        port->printf("| %-16s | %i |\n", s, (int)((AP_Int8 *)ap)->get());
```
don't need the pretty printer any more

andyp1per replied:
Removed

tridge commented on libraries/AP_Param/AP_Param.cpp:
```diff
@@ -2551,16 +2571,29 @@ void AP_Param::show_all(AP_HAL::BetterStream *port, bool showKeyValues)
     enum ap_var_type type;
+    const bool hide_disable = _hide_disabled_groups;
+
+    _hide_disabled_groups = false;
```
can be removed

andyp1per replied:
Removed

tridge reviewed (changes requested):
need to get an idea of flash cost, so we can decide if we need this to be able to be configured on/off

andyp1per commented:
```
Target          Text    Data  BSS     Total  
---------------------------------------------
bin/arducopter  900496  2644  259576  1162716

Target         Text    Data  BSS     Total  
--------------------------------------------
bin/arduplane  901740  2560  259660  1163960
```
without
```
Target          Text    Data  BSS     Total  
---------------------------------------------
bin/arducopter  893452  1324  260900  1155676

Target         Text    Data  BSS     Total  
--------------------------------------------
bin/arduplane  894992  1320  260896  1157208
```

So 7k flash on copter 6.7k on plane

andyp1per commented:
The main target of this will be 1mb boards since they are the ones that generally have OSD, I don't see how we can turn this off by default if its to be at all useful.

andyp1per commented:
@vierfuffzig getting close to a mergeable state - would be great iff you could give it another whirl!

vierfuffzig commented:
@andyp1per thanks again, awesome progress!
i did quite a bit of testing, things working great so far.
when i updated from previous version i actually had a couple of IMU temp resets on fifo errors, like those on WingFC10 here: https://github.com/ArduPilot/ardupilot/issues/15090

```
param set SERIAL2_BAUD 57
LOITER> MPU: temp reset IMU[0] -624 1744
MPU: IMU[0] stop at 24 of 73
MPU: temp reset IMU[0] -640 -4353
MPU: IMU[0] stop at 24 of 74
param show MPU: MPU60x0: error in fifo read 112 bytes

MPU: temp reset IMU[0] -592 1789
MPU: IMU[0] stop at 24 of 73
MPU: temp reset IMU[0] -592 1542
MPU: IMU[0] stop at 24 of 74
```

memory issues due to deprecated / invalid previous settings? i could not see or reproduce those again. testing done on a matekF405-wing, ~17k freemem left btw.

i did a full reset and fresh param screen setup, working as expected.
it might add convenience to have defaults for each param's / save_button's _x and _y to avoid having to set each of them individually:
param's x = 2; y = 2 - 9
save x = 20; y = 12
to make everything fit on smallest common screen format.

will do more testing after recharging my batteries ;)

andyp1per commented:
@vierfuffzig - apologies, that's actually a bug - hopefully fixed now

vierfuffzig commented:
@andyp1per works, thanks! PRed _x and _y defaults to make sure everything is visible in NTSC format too.

i do see one issue: if you set a param slot with a type, this type setting will persist through selecting a different param of a different type on that slot via stick input. so if i had SERIALn_PROTOCOL with respective type set in slot 1 and then pick SERIALn_BAUD for that slot via stick input, i'll still have the SERIAL_PROTOCOL options to choose from. if not aware of that, i can perfectly set that param away from any useable value, but not back. is it possible to reset "type" to none if a different parameter is selected in a slot that had a none-NONE "type"setting before? this would likely help to keep trouble away.

andyp1per commented:
@vierfuffzig I think I have fixed that bug.

Not sure about the NTSC thing, how many lines do you get?

vierfuffzig commented:
wt...?? that was 12 minutes only for that fix...
will try.

13 visible lines on NTSC (0 - 12), borders might get cut on sub par centering

vierfuffzig commented:
> @vierfuffzig I think I have fixed that bug

working, awesome!

andyp1per commented:
I have added a 9th parameter and shuffled them up to fit on an NTSC screen. I've also changed the defaults to be what I think are a reasonable set. This is the result:

![image](https://user-images.githubusercontent.com/2893260/91661594-ef7af300-ead4-11ea-96b4-6a281a3d1010.png)

I think this is done now

andyp1per commented:
The extra parameters adds 90 bytes, so no material difference

andyp1per commented:
@vierfuffzig and testing really appreciated!

vierfuffzig commented:
@andyp1per working as expected, nice additions!
just noticed plane AUX_FUNCTION needs alignment with current options.
not sure if implementing all options in the 200 / 300 range makes any sense?

andyp1per commented:
Added support for build time disabling.
Stack left when in OSD menu is 712 bytes

andyp1per commented:
Re-based on @yaapu 's PR

peterbarker commented on libraries/AP_OSD/AP_OSD_ParamScreen.cpp:
```diff
@@ -0,0 +1,710 @@
+        // there is no way to validate the ranges, so as a rough guess prevent
+        // integer types going below -1;
+        case AP_PARAM_INT8: 
+        {
```
Atypical formatting; usually the break goes inside the {} and the { would be on the same line as the colon on the case statement.

andyp1per replied:
Fixed - and all the other places as well

tridge commented:
when building boards without a SPI OSD, and without this new param feature, it is no longer eliding the draw methods from the AP_OSD_Screen class, which means it costs us an exra 9k of flash.
example:
 - build master CubeOrange with --osd
 - build PR CubeOrange with --osd but with this feature disabled
 - flash size goes up by 9k

andyp1per commented:
Build of master with CubOrange:
```
Build directory: /mnt/i/github/ardupilot-master/build/CubeOrange
Target          Text     Data  BSS     Total
----------------------------------------------
bin/arducopter  1481972  1520  129764  1613256
```
This PR:
```
Build directory: /mnt/i/github/ardupilot-dev/build/CubeOrange
Target          Text     Data  BSS     Total
----------------------------------------------
bin/arducopter  1482384  1520  129760  1613664
```
So about 400bytes difference

andyp1per commented:
@tridge 

- [X] build on cube orange does not change significantly in size
- [X] configure_all.py can build all the boards - I had to modify KakuteF7 and OMNIBUSF7V2 to fit, but I think these changes should have been in there anyway, also unrelated to my change f103-Trigger has a build error - perhaps from a recent merge?
- [X] addressed Peter's formatting changes
- [X] re-tested
- [X] ready to merge

tridge commented:
this is good to merge when CI passes

peterbarker commented:
Merged, thanks!

drtrigon commented:
Is this in stable already? I cannot find `OSD5_ENABLE` or `OSD6_ENABLE` params in ArduCopter 4.0.7.

khancyr commented:
It should come with the v4.1 serie

drtrigon commented:
Too bad! ;) Thanks for the info!
